conda/gateways/repodata/jlap/fetch.py

```diff
@@ -0,0 +1,402 @@
+            )
+            apply.append(patch)
+            want = patch["from"]
+
+    if have != want:
+        log.debug(f"No patch from local revision {format_hash(have)}")
+        raise JlapPatchNotFound(f"No patch from local revision {format_hash(have)}")
+
+    return apply
+
+
+def apply_patches(data, apply):
+    while apply:
+        patch = apply.pop()
+        log.debug(
+            f"{format_hash(patch['from'])} \N{RIGHTWARDS ARROW} {format_hash(patch['to'])}, "
+            f"{len(patch['patch'])} steps"
+        )
+        data = jsonpatch.JsonPatch(patch["patch"]).apply(data, in_place=True)
+
+
+def withext(url, ext):
+    return re.sub(r"(\.\w+)$", ext, url)
+
+
+@contextmanager
+def timeme(message):
+    begin = time.monotonic()
+    yield
+    end = time.monotonic()
+    log.debug("%sTook %0.02fs", message, end - begin)
+
+
+def build_headers(json_path: pathlib.Path, state: RepodataState):
+    """
+    Caching headers for a path and state.
+    """
+    headers = {}
+    # simplify if we require state to be empty when json_path is missing.
+    if json_path.exists():
+        etag = state.get("_etag")
+        if etag:
+            headers["if-none-match"] = etag
+    return headers
+
+
+class HashWriter(io.RawIOBase):
+    def __init__(self, backing, hasher):
+        self.backing = backing
+        self.hasher = hasher
+
+    def write(self, b: bytes):
+        self.hasher.update(b)
+        return self.backing.write(b)
+
+    def close(self):
+        self.backing.close()
+
+
+def download_and_hash(
+    hasher, url, json_path, session: Session, state: RepodataState | None, is_zst=False
+):
+    """
+    Download url if it doesn't exist, passing bytes through hasher.update()
+    """
+    state = state or RepodataState()
+    headers = build_headers(json_path, state)
+    timeout = context.remote_connect_timeout_secs, context.remote_read_timeout_secs
+    response = session.get(url, stream=True, timeout=timeout, headers=headers)
+    log.debug("%s %s", url, response.headers)
+    response.raise_for_status()
+    length = 0
+    # is there a status code for which we must clear the file?
+    if response.status_code == 200:
+        if is_zst:
+            decompressor = zstandard.ZstdDecompressor()
+            writer = decompressor.stream_writer(
+                HashWriter(json_path.open("wb"), hasher), closefd=True  # type: ignore
+            )
+        else:
+            writer = HashWriter(json_path.open("wb"), hasher)
+        with writer as repodata:
+            for block in response.iter_content(chunk_size=1 << 14):
+                repodata.write(block)
+    if response.request:
+        log.info("Download %d bytes %r", length, response.request.headers)
+    return response  # can be 304 not modified
+
+
+def request_url_jlap_state(
+    url, state: RepodataState, get_place=get_place, full_download=False, *, session: Session
+):
```

Comment:
This is a sufficiently complex function and needs to be split up into smaller units

Reply:
I would like to reformulate it as a formal state machine.

Reply:
@kenodegard would you like to take on the refactor for this function, submitting it back as a PR on top of this one? It should be easier now that there is only one download-and-hash(is_zst=True/False) function.